### PR TITLE
Support for docker secret

### PR DESCRIPTION
**Pull request by bdpiparva**
* Allows user to bind existing docker secret with containers

Screenshot

screen shot 2017-09-06 at 4 52 26 pm

### Usage

1. Plugin requires source of secret to bind with container. 

	```
	src=NAME_OF_THE_SECRET
	```

      This will be available at default mounting point `/run/secrets/NAME_OF_THE_SECRET` with `uid=0`, `gid=0` and permission `-r--r--r--`

2. Bind secret at specific location on container 

	```
	src=NAME_OF_THE_SECRET, target=/tmp/secrets/NAME_OF_THE_SECRET
	```

      This will mount secret at `/tmp/secrets/` with filename `NAME_OF_THE_SECRET`

3. Other options

	```
	src=NAME_OF_THE_SECRET, target=/tmp/secrets/NAME_OF_THE_SECRET, uid=1001,  gid=1002, mode=0640L
	```

### Properties 

Property     | Desciption
-------------|------------------
src          | Name of the secret
target       | Mount location for secret `Default location: /run/secrets/`
uid          | user id `Default :0`
gid          | group id `Default: 0`
mode         | permission for secret file `Default: -r--r--r--`

### CODE REVIEW

File: src/main/java/cd/go/contrib/elasticagents/dockerswarm/elasticagent/DockerSecret.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
+package cd.go.contrib.elasticagents.dockerswarm.elasticagent;
+
+import com.spotify.docker.client.messages.swarm.SecretSpec;
+import org.glassfish.jersey.internal.util.Base64;
```
Comment (ketan) [inline src/main/java/cd/go/contrib/elasticagents/dockerswarm/elasticagent/DockerSecret.java]: `Base64` is provided by `java.util.Base64`.

File: src/main/java/cd/go/contrib/elasticagents/dockerswarm/elasticagent/DockerService.java [hunk 2]
```diff
@@ -69,55 +70,102 @@ public void terminate(DockerClient docker) throws DockerException, InterruptedEx
...
             LOG.warn("Cannot terminate a service that does not exist " + name);
         }
     }
 
+    private static void deleteSecrets(DockerClient docker, String serviceName) throws DockerException, InterruptedException {
+        docker.listSecrets().forEach(secret -> {
+            if (serviceName.equals(secret.secretSpec().labels().get(SWARM_SERVICE_NAME))) {
+                try {
+                    LOG.debug("Deleting secret " + secret.secretSpec().name());
+                    docker.deleteSecret(secret.id());
```
Comment (ketan) [inline src/main/java/cd/go/contrib/elasticagents/dockerswarm/elasticagent/DockerService.java]: Why are we deleting secrets? Isn't the point to just run `docker service create --secret src=SECRET_NAME,target=/path/to/file/in/container` and use an existing secret?

Won't this cause issues the next time some service wants to use the same secret?

Comment (bdpiparva) [reply src/main/java/cd/go/contrib/elasticagents/dockerswarm/elasticagent/DockerService.java]: What if there is a change in secret, or deleted from profile ?

Comment (ketan) [reply src/main/java/cd/go/contrib/elasticagents/dockerswarm/elasticagent/DockerService.java]: The intention is that secrets will be provided outside of the plugin. So admins will be expected to upload the secrets out-of-band from the CLI or something similar. I'm not sure how you could possibly provide a plethora of secrets which can be several kilobytes in a profile in the cruise-config.xml. Some secrets will likely be shared between profiles so will cause duplication errors.

**Comment by arvindsv** [comment]
Yes, I would expect that they are managed externally - with some way of accessing them from within GoCD.

File: src/main/java/cd/go/contrib/elasticagents/dockerswarm/elasticagent/DockerService.java (hunk 2 above)
Comment (ketan) [reply src/main/java/cd/go/contrib/elasticagents/dockerswarm/elasticagent/DockerService.java]: Remember that binary blobs encoded as base64 will add about 33% to the size of the binary blob and add to the size of the config file.

**Comment by arvindsv** [comment]
Or a plugin, as in this case. So, maybe the name of a secret can be specified here, and it can be used during startup of a container?

**Comment by bdpiparva** [comment]
Profile Validations:

----------------------
screen shot 2017-09-06 at 4 52 26 pm

----------------------
screen shot 2017-09-07 at 10 35 14 am

----------------------
screen shot 2017-09-07 at 10 35 42 am

**Comment by ketan** [comment]
Can you add a test to ascertain mode when leading 0 is removed? It is likely that 640 will be interpreted as an integer instead of long

**Comment by bdpiparva** [comment]
sure, will put a test to validate octal

**Pull request merged**